File: module-server/src/main/java/de/twomartens/wahlrecht/mapper/v1/ElectionResultMapper.java
package de.twomartens.wahlrecht.mapper.v1;

import de.twomartens.wahlrecht.model.internal.ElectionResult;
import de.twomartens.wahlrecht.model.internal.VotingResult;
import java.util.Collection;
import java.util.Map;
import org.mapstruct.CollectionMappingStrategy;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.ReportingPolicy;

@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface ElectionResultMapper {

  de.twomartens.wahlrecht.model.dto.v1.ElectionResult mapToExternal(de.twomartens.wahlrecht.model.db.ElectionResult result);

  Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult> mapToExternal(
      Collection<de.twomartens.wahlrecht.model.db.VotingResult> results);

  Map<Integer, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>> mapToExternal(
      Map<Integer, Collection<de.twomartens.wahlrecht.model.db.VotingResult>> results);

  @Mapping(target = "id", ignore = true)
  @Mapping(target = "created", ignore = true)
  @Mapping(target = "lastModified", ignore = true)
  de.twomartens.wahlrecht.model.db.ElectionResult mapToDB(de.twomartens.wahlrecht.model.dto.v1.ElectionResult result);

  Collection<de.twomartens.wahlrecht.model.db.VotingResult> mapToDB(
      Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult> results);

  Map<Integer, Collection<de.twomartens.wahlrecht.model.db.VotingResult>> mapToDB(
      Map<Integer, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>> results);

  ElectionResult mapToInternal(de.twomartens.wahlrecht.model.dto.v1.ElectionResult result);

  Collection<VotingResult> mapToInternal(
      Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult> results);

  Map<Integer, Collection<VotingResult>> mapToInternal(
      Map<Integer, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>> results);
}

File: module-server/src/main/kotlin/de/twomartens/wahlrecht/mapper/v1/ElectionResultMapper.kt
package de.twomartens.wahlrecht.mapper.v1

import de.twomartens.wahlrecht.model.db.ElectionResult
import de.twomartens.wahlrecht.model.db.VotingResult
import org.mapstruct.*
import org.mapstruct.factory.Mappers

@Mapper(
    collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
    unmappedTargetPolicy = ReportingPolicy.IGNORE
)
interface ElectionResultMapper {
    companion object {
        val VOTING_RESULT_MAPPER: VotingResultMapper = Mappers.getMapper(VotingResultMapper::class.java)
    }

    fun mapToExternal(result: ElectionResult): de.twomartens.wahlrecht.model.dto.v1.ElectionResult

    fun mapToExternal(results: Collection<VotingResult>): Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>

    fun mapToExternal(results: Map<Int, Collection<VotingResult>>):
            Map<Int, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>>

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "created", ignore = true)
    @Mapping(target = "lastModified", ignore = true)
    fun mapToDB(result: de.twomartens.wahlrecht.model.dto.v1.ElectionResult): ElectionResult

    fun mapToDB(results: Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>): Collection<VotingResult>

    fun mapToDB(results: Map<Int, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>>):
            Map<Int, Collection<VotingResult>>

    fun mapToInternal(result: de.twomartens.wahlrecht.model.dto.v1.ElectionResult):
            de.twomartens.wahlrecht.model.internal.ElectionResult

    fun mapToInternal(result: de.twomartens.wahlrecht.model.dto.v1.VotingResult):
            de.twomartens.wahlrecht.model.internal.VotingResult {
        return VOTING_RESULT_MAPPER.mapToInternal(result)
    }

    fun mapToInternal(results: Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>):
            Collection<de.twomartens.wahlrecht.model.internal.VotingResult>

    fun mapToInternal(results: Map<Int, Collection<de.twomartens.wahlrecht.model.dto.v1.VotingResult>>):
            Map<Int, Collection<de.twomartens.wahlrecht.model.internal.VotingResult>>
}